visualization/app/codeCharta/state/store/metricData/nodeMetricData/nodeMetricData.service.ts

```diff
@@ -0,0 +1,76 @@
+import { StoreService, StoreSubscriber } from "../../../store.service"
+import { IRootScopeService } from "angular"
+import { NodeMetricDataActions, calculateNewNodeMetricData } from "./nodeMetricData.actions"
+import { isActionOfType } from "../../../../util/reduxHelper"
+import { AttributeTypeValue, BlacklistItem, NodeMetricData } from "../../../../codeCharta.model"
+import { FilesSelectionSubscriber, FilesService } from "../../files/files.service"
+import { BlacklistService, BlacklistSubscriber } from "../../fileSettings/blacklist/blacklist.service"
+import { AttributeTypesService, AttributeTypesSubscriber } from "../../fileSettings/attributeTypes/attributeTypes.service"
+import { FileState } from "../../../../model/files/files"
+
+export interface NodeMetricDataSubscriber {
+	onNodeMetricDataChanged(nodeMetricData: NodeMetricData[])
+}
+
+export class NodeMetricDataService implements StoreSubscriber, FilesSelectionSubscriber, BlacklistSubscriber, AttributeTypesSubscriber {
+	public static UNARY_METRIC = "unary"
+	private static NODE_METRIC_DATA_CHANGED_EVENT = "node-metric-data-changed"
+
+	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
+		StoreService.subscribe(this.$rootScope, this)
+		FilesService.subscribe(this.$rootScope, this)
+		BlacklistService.subscribe(this.$rootScope, this)
+		AttributeTypesService.subscribe(this.$rootScope, this)
+	}
+
+	public onStoreChanged(actionType: string) {
+		if (isActionOfType(actionType, NodeMetricDataActions)) {
+			this.notify(this.select())
+		}
+	}
+
+	public onFilesSelectionChanged(files: FileState[]) {
+		this.storeService.dispatch(calculateNewNodeMetricData(files, this.storeService.getState().fileSettings.blacklist))
+	}
+
+	public onBlacklistChanged(blacklist: BlacklistItem[]) {
+		this.storeService.dispatch(calculateNewNodeMetricData(this.storeService.getState().files, blacklist))
+	}
+
+	public onAttributeTypesChanged() {
+		this.storeService.dispatch(
+			calculateNewNodeMetricData(this.storeService.getState().files, this.storeService.getState().fileSettings.blacklist)
+		)
+	}
+
+	public getMetrics(): string[] {
+		return this.storeService.getState().metricData.nodeMetricData.map(x => x.name)
+	}
+
+	public isMetricAvailable(metricName: string): boolean {
+		return this.storeService.getState().metricData.nodeMetricData.some(x => x.name == metricName)
+	}
+
+	public getMaxMetricByMetricName(metricName: string): number {
```

Comment:
```suggestion
	public getMaxMetricByMetricName(metricName: string): number | undefined {
```

I could imagine that metric is always truthy due to just accepting known metricNames (I suggest to use an enum to guarantee that metricName is always known).

Reply:
Yes, we only call this method on metrics we know exist.

How can we build an enum dynamically containing only the metrics that are relevant to the visible files though?